File: src/java/org/orbeon/oxf/servlet/OrbeonServletFilter.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.servlet;

import org.orbeon.oxf.pipeline.api.WebAppExternalContext;
import org.orbeon.oxf.webapp.OrbeonClassLoader;

import javax.servlet.*;
import java.io.IOException;

/**
 * OXFServletFilter is the Servlet Filter entry point of OXF.
 *
 * WARNING: This class must only depend on the Servlet API and the OXF Class Loader.
 */
public class OrbeonServletFilter implements Filter {

    // Servlet Filter delegate
    private Filter servletFilterDelegate;

    // WebAppExternalContext
    private WebAppExternalContext webAppExternalContext;

    public void init(FilterConfig config) throws ServletException {
        try {
            // Instanciate WebAppExternalContext
            webAppExternalContext = new ServletWebAppExternalContext(config.getServletContext());

            // Instanciate Servlet Filter delegate
            Class delegateServletClass = OrbeonClassLoader.getClassLoader(webAppExternalContext).loadClass(OrbeonServletFilter.class.getName() + OrbeonClassLoader.DELEGATE_CLASS_SUFFIX);
            servletFilterDelegate = (Filter) delegateServletClass.newInstance();

            // Initialize Servlet Filter delegate
            Thread currentThread = Thread.currentThread();
            ClassLoader oldThreadContextClassLoader = currentThread.getContextClassLoader();
            try {
                currentThread.setContextClassLoader(OrbeonClassLoader.getClassLoader(webAppExternalContext));
                servletFilterDelegate.init(config);
            } finally {
                currentThread.setContextClassLoader(oldThreadContextClassLoader);
            }
        } catch (Exception e) {
            throw new ServletException(e);
        }
    }

    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
        // Delegate to Servlet Filter delegate
        Thread currentThread = Thread.currentThread();
        ClassLoader oldThreadContextClassLoader = currentThread.getContextClassLoader();
        try {
            currentThread.setContextClassLoader(OrbeonClassLoader.getClassLoader(webAppExternalContext));
            servletFilterDelegate.doFilter(request, response, chain);
        } finally {
            currentThread.setContextClassLoader(oldThreadContextClassLoader);
        }
    }

    public void destroy() {
        // Delegate to Servlet Filter delegate
        Thread currentThread = Thread.currentThread();
        ClassLoader oldThreadContextClassLoader = currentThread.getContextClassLoader();
        try {
            currentThread.setContextClassLoader(OrbeonClassLoader.getClassLoader(webAppExternalContext));
            servletFilterDelegate.destroy();
        } finally {
            currentThread.setContextClassLoader(oldThreadContextClassLoader);
        }
    }
}

File: src/main/scala/org/orbeon/oxf/servlet/OrbeonServletFilter.scala
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.servlet

import javax.servlet._
import javax.servlet.http.{HttpServletRequest, HttpServletResponse}

import org.orbeon.oxf.externalcontext.WebAppContext
import org.orbeon.oxf.pipeline.api._
import org.orbeon.oxf.processor.ServletFilterGenerator
import org.orbeon.oxf.webapp.ServletPortlet._
import org.orbeon.oxf.webapp.{ProcessorService, ServletPortlet}

import scala.collection.JavaConverters._

// For backward compatibility
class OrbeonServletFilterDelegate extends OrbeonServletFilter

class OrbeonServletFilter extends Filter with ServletPortlet {

  private implicit val logger = ProcessorService.Logger

  def logPrefix = "Servlet filter"

  // Immutable map of servlet parameters
  private var _initParameters: Map[String, String] = _
  def initParameters = _initParameters

  // Filter init
  def init(config: FilterConfig): Unit =
    withRootException("initialization", new ServletException(_)) {

      _initParameters =
        config.getInitParameterNames.asScala map
          (n ⇒ n → config.getInitParameter(n)) toMap

      init(WebAppContext(config.getServletContext), None)
    }

  // Filter destroy
  def destroy(): Unit =
    withRootException("destruction", new ServletException(_)) {
      destroy(None)
    }

  // Filter request
  def doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain): Unit =
    withRootException("request", new ServletException(_)) {
      val pipelineContext = new PipelineContext
      pipelineContext.setAttribute(ServletFilterGenerator.FILTER_CHAIN, chain)
      val externalContext = new ServletExternalContext(pipelineContext, webAppContext, request.asInstanceOf[HttpServletRequest], response.asInstanceOf[HttpServletResponse])
      processorService.service(pipelineContext, externalContext)
    }
}